vl53l0x: Add support for interrupt and configurable sampling frequency

titouanc opened the pull request:
- drivers: sensor: vl53l0x: run clang-format
- drivers: sensor: vl53l0x: add attr_get/set for the sampling frequency
- drivers: sensor: vl53l0x: add support for interrupt-gpios
- boards: shields: x_nucleo_53l0a1: add interrupt-gpios
- samples: sensors: vl53l0x: update to use trigger mode if available

titouanc committed:
drivers: sensor: vl53l0x: add attr_get/set for the sampling frequency

Add implementation for sensor_attr_get/set to configure the sensor sampling
frequency (ie inverse of the timing budget for a measurement)

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
drivers: sensor: vl53l0x: run clang-format

Reformat the file to the latest code formatting standard with clang-format
before making new changes

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
drivers: sensor: vl53l0x: add support for interrupt-gpios

- Add a new optional device tree property `interrupt-gpios`
- When this property is defined, sensor_sample_fetch() now waits for an
  interrupt from the sensor, rather than actively polling for completion
- Add implementation for sensor_trigger_set()

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
boards: shields: x_nucleo_53l0a1: add interrupt-gpios

Add the new device-tree property `interrupt-gpios` for the sensor soldered
on the shield, and add examples of other possible pinouts as examples. The
alternate pinouts can be configured with solder drops on the shield

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
samples: sensors: vl53l0x: update to use trigger mode if available

The sample now demonstrates how to configure the sampling frequency, and
how to gather data using the "new data ready" trigger. If this is not
supported (ie the sensor has no interrupt pin wired), fallback to polling
mode as before.

Also reformat the sample with clang-format to adhere to the latest code
format standard.

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

MaureenHelm commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -263,9 +263,79 @@ static int vl53l0x_channel_get(const struct device *dev, enum sensor_channel cha
+	struct vl53l0x_data *drv_data = dev->data;
+	int ret = VL53L0X_GetMeasurementTimingBudgetMicroSeconds(&drv_data->vl53l0x, &timing);
+	if (ret) {
+		LOG_ERR("[%s] VL53L0X_GetMeasurementTimingBudgetMicroSeconds failed", dev->name);
```
Please use more meaningful error messages instead of just printing painfully long function names.

titouanc replied:
I agree with that. However all the existing log messages in this driver (since its birth in Zephyr https://github.com/zephyrproject-rtos/zephyr/commit/347560459b0c55efe95e2764197a96ce22247b02) use this same form, therefore we would have to change them all for consistency ?

MaureenHelm replied:
If you're up for that yes, but I'd also be fine with just making your new log messages meaningful even if they're inconsistent with the old ones.

MaureenHelm commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -263,9 +263,79 @@ static int vl53l0x_channel_get(const struct device *dev, enum sensor_channel cha
+		LOG_ERR("[%s] VL53L0X_GetMeasurementTimingBudgetMicroSeconds failed", dev->name);
+		return ret;
+	}
+	val->val1 = 1000000 / timing;
+	val->val2 = 1000000 / (1000000 % timing) / timing;
```
Can timing ever be zero?

MaureenHelm commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -263,9 +263,79 @@ static int vl53l0x_channel_get(const struct device *dev, enum sensor_channel cha
+
+static int vl53l0x_set_sampling_freq(const struct device *dev, const struct sensor_value *val)
+{
+	uint64_t timing = 1000000000000 / (1000000 * val->val1 + val->val2);
```
Use `USEC_PER_SEC` and friends

MaureenHelm commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -42,16 +43,77 @@ LOG_MODULE_REGISTER(VL53L0X, CONFIG_SENSOR_LOG_LEVEL);
 
 struct vl53l0x_data {
+	const struct device *dev;
 	bool started;
 	VL53L0X_Dev_t vl53l0x;
-	VL53L0X_RangingMeasurementData_t RangingMeasurementData;
+	VL53L0X_RangingMeasurementData_t measurement;
```
This change doesn't look related to interrupt-gpios. Please extract to a separate commit.

MaureenHelm commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -42,16 +43,77 @@ LOG_MODULE_REGISTER(VL53L0X, CONFIG_SENSOR_LOG_LEVEL);
 };
 
-static int vl53l0x_setup_single_shot(const struct device *dev)
+static inline bool vl53l0x_supports_chan(enum sensor_channel chan)
+{
+	switch (chan) {
+	case SENSOR_CHAN_DISTANCE:
+	case SENSOR_CHAN_PROX:
+	case SENSOR_CHAN_ALL:
+		return true;
+	default:
+		return false;
+	}
+}
```
This change doesn't look related to interrupt-gpios. Please extract to a separate commit.

MaureenHelm commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -140,6 +196,18 @@ static int vl53l0x_setup_single_shot(const struct device *dev)
 	return ret;
 }
 
+static int vl53l0x_set_ranging_mode(const struct device *dev, VL53L0X_DeviceModes mode)
+{
+	struct vl53l0x_data *drv_data = dev->data;
+	int ret = VL53L0X_SetDeviceMode(&drv_data->vl53l0x, mode);
+	if (ret) {
+		LOG_ERR("[%s] VL53L0X_SetDeviceMode failed", dev->name);
+		return ret;
+	}
+	drv_data->current_mode = mode;
+	return 0;
+}
```
This change doesn't look related to interrupt-gpios. Please extract to a separate commit.

titouanc replied:
In my opinion, this change is related to interrupt gpios. Before this commit, the sensor was always in single-shot mode (ie the i2c master tells the sensor to start a measurement, waits, then fetch the result from the sensor).

In this commit, we add support for `sensor_trigger_set` for `SENSOR_TRIG_DATA_READY` in the vl53l0x driver. For this trigger to make sense, we need to put the sensor in continuous mode (ie the sensor continuously makes new measurements, and notifies the i2c master via an interrupt that a new measurement is ready).

In continuous mode, the caller does not need to start a new measurement in `sensor_sample_fetch`, so we need to track the current mode of the device to know if a new measurement should be started. This function simply wraps the call to the ST HAL, so that the current device mode is properly tracked. 

Finally, the device mode is not exposed to the user, so this function is only an accompanying helper for the interrupt gpio functionality

titouanc committed:
drivers: sensor: vl53l0x: Rename driver data member for consistency

Rename the member `RangingMeasurementData` as `measurement` in the struct
`vl53l0x_data` (runtime driver data) for consistency with the other members
and to improve overall codestyle

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc committed:
drivers: sensor: vl53l0x: Return -ENOTSUP for unsupported channels

Up to now, calling vl53l0x_sample_fetch() with an unsupported channel
resulted in a fatal system error. However, it is more appropriate to
return -ENOTSUP in this case.

Also introduce the helper function vl53l0x_supports_chan which will
be helpful in subsequent commits.

Signed-off-by: Titouan Christophe <moiandme@gmail.com>

titouanc commented:
Thank you for reviewing this PR @MaureenHelm 

I applied your suggestions, fixed some issues uncovered by the compliance checks, and rebased on top of the latest `main`

titouanc commented:
I clicked on "Re-request review" for @MaureenHelm and this removed all the other reviewers, not sure why :thinking:

jfischer-no commented on drivers/sensor/vl53l0x/vl53l0x.c:
```diff
@@ -62,52 +62,40 @@ static int vl53l0x_setup_single_shot(const struct device *dev)
 
 	ret = VL53L0X_StaticInit(&drv_data->vl53l0x);
 	if (ret) {
-		LOG_ERR("[%s] VL53L0X_StaticInit failed",
```
Don't, unnecessary changes, https://github.com/zephyrproject-rtos/zephyr/issues/48785
